### ISSUE TITLE
[no-useless-constructor] `constructor` with `super(someArgs)` is not useless in `Angular`

### ISSUE THREAD

**Issue by JounQin** [typescript-eslint/typescript-eslint#1164]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-useless-constructor": 2
  }
}
```

```TS
class A {
  protected http: HttpClient
  constructor(protected injector: Injector) {
    this.http = this.injector.get(HttpClient)
  }
}

class B implements A {
  // this should not be considered as useless due to DI
  constructor(injector: Injector) {
    super(injector)
  }
}
```

**Expected Result**

No error report

**Actual Result**

Error

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.6.0` |
| `@typescript-eslint/parser`        | `2.6.0` |
| `TypeScript`                       | `3.6.4` |
| `ESLint`                           | `6.6.0` |
| `node`                             | `13.0.1` |
| `npm`                              | `6.12.0` |

**Comment by JounQin** [comment]
We can use following workaround temporarily:

```ts
class B implments A {
  constructor(protected injector: Injector) {
    super(injector);
  }
}
```

But if `A` declare it as `private`, we can only change it to another property name like:

```ts
class A {
  private http: HttpClient
  constructor(private injector: Injector) {
    this.http = this.injector.get(HttpClient)
  }
}

class B implements A {
  constructor(private _injector: Injector) {
    super(_injector)
  }
}
```

**Comment by bradzacher** [comment]
Could you please explain why you need the constructor on the child class?

Unless I'm mistaken, this is perfectly valid code, which does exactly the same thing, without the useless constructor:
```ts
class A {
  protected http: HttpClient
  constructor(protected injector: Injector) {
    this.http = this.injector.get(HttpClient)
  }
}

class B extends A { }
```

**Comment by JounQin** [comment]
Oh sorry, it should be `extends`, it's a typo.

B's constructor is required in favor of `DI` system in Angular.

**Comment by bradzacher** [comment]
Have you got some docs on why the constructor is explicitly needed on the child?
I haven't used angular since like 1.6, so I don't know how it works.

Unless I'm mistaken it shouldn't need it explicitly, as the child should implicitly inherit the parent's constructor.

**Comment by JounQin** [comment]
@bradleyayers Just take <https://stackblitz.com/edit/angular-gh3sf4> as an example.

**Comment by bradzacher** [comment]
Please be careful with your @ tagging. You tagged the wrong account.
Github has a crappy implementation, and doesn't put my name first in the autocomplete list.

---

Looking into this, I have no idea why it's required. I think angular statically analyses the file by itself? I'm not sure, the custom  compiler is a black box to me.

We don't want to encode the semantics of angular into typescript.
The fact that the constructor is required by the angular compiler is an implementation detail of the framework.

If you'd like this functionality, I'd suggest getting a rule created in a plugin like eslint-plugin-angular.

**Comment by JounQin** [comment]
@bradzacher Thx for your patients. I will open a feature request issue at [angular-eslint](https://github.com/angular-eslint/angular-eslint) then.
